sdk/go/common/encoding/marshal.go

```diff
@@ -118,3 +124,65 @@ func (m *yamlMarshaler) Unmarshal(data []byte, v interface{}) error {
 	}
 	return nil
 }
+
+type gzipMarshaller struct {
+	inner Marshaler
+}
+
+func (m *gzipMarshaller) IsJSONLike() bool {
+	return m.inner.IsJSONLike()
+}
+
+func (m *gzipMarshaller) IsYAMLLike() bool {
+	return m.inner.IsYAMLLike()
+}
+
+func (m *gzipMarshaller) Marshal(v interface{}) ([]byte, error) {
+	b, err := m.inner.Marshal(v)
+	if err != nil {
+		return nil, err
+	}
+
+	var buf bytes.Buffer
+	writer := gzip.NewWriter(&buf)
+	defer writer.Close()
+	_, err = writer.Write(b)
+	if err != nil {
+		return nil, err
+	}
+	if err := writer.Close(); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+
+}
+
+func (m *gzipMarshaller) Unmarshal(data []byte, v interface{}) error {
+	buf := bytes.NewBuffer(data)
+	reader, err := gzip.NewReader(buf)
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+	inflated, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return err
+	}
+	if err := reader.Close(); err != nil {
+		return err
+	}
+	return m.inner.Unmarshal(inflated, v)
+}
+
+// IsCompressed returns if data is zip compressed.
+func IsCompressed(data []byte) bool {
+	return data[0] == 31 && data[1] == 139
```

Comment:
Ah! This is very interesting.

Is there a lib function we can use?

Does this panic on `len(data)< 2`?

Reply:
There is a library function we can repurpose, but will technically violate its contract. This is the check that `compress/gzip` uses internally (constants translated). 
```go
// IsCompressed returns if data is zip compressed.
func IsCompressed(buf []byte) bool {
	// Taken from compress/gzip/gunzip.go
	return len(buf) >= 3 && buf[0] == 31 && buf[1] == 139 && buf[2] == 8
}
```

It is a superset of the previous check.